src/Proto.Mailbox/NonBlockingBoundedMailbox.cs

```diff
@@ -0,0 +1,48 @@
+﻿// -----------------------------------------------------------------------
+//   <copyright file="NonBlockingBoundedMailbox.cs" company="Asynkron HB">
+//       Copyright (C) 2015-2017 Asynkron HB All rights reserved
+//   </copyright>
+// -----------------------------------------------------------------------
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Proto.Mailbox
+{
+    public class NonBlockingBoundedMailbox : IMailboxQueue
+    {
+        public NonBlockingBoundedMailbox(int maxSize, Action<object> overflowAction, TimeSpan timeout)
+        {
+            _maxSize = maxSize;
+            _overflowAction = overflowAction;
+            _timeout = timeout;
+        }
+
+        private readonly ConcurrentQueue<object> _messages = new ConcurrentQueue<object>();
+        private readonly int _maxSize;
+        private readonly Action<object> _overflowAction;
+        private readonly TimeSpan _timeout;
+
+        public void Push(object message)
+        {
+            if (SpinWait.SpinUntil(() => _messages.Count < _maxSize, _timeout))
+            {
+                //this will be racy, but best effort is good enough..
+                _messages.Enqueue(message);
+            }
```

Comment:
Once the condition is met, another thread could have enqueued a message on the queue and the condition is now false again. but best effort is good enough here. 
the mail purpose is to prevent the system to tip over due to over load. 1-2 messages too many on the queue will not cause that.